datafusion/expr-common/src/columnar_value.rs

```diff
@@ -195,6 +214,10 @@ impl ColumnarValue {
                 kernels::cast::cast_with_options(array, cast_type, &cast_options)?,
             )),
             ColumnarValue::Scalar(scalar) => {
+                if scalar.data_type().logically_eq(cast_type) {
```

Comment:
This function is also used in physical expr `CastExpr`. But we should not do logical equality check for physical casting, we do actual arrow cast instead. I think it makes sense to have 2 casting function, one do logical eq check, one do actual casting

Reply:
I think the general idea is we do logical type eq check in logical layer. Do actual casting in physical layer.

Reply:
Seems like ColumnarValue is used in both logical and physical layer so I'm not really sure how to approach this.

Reply:
> I think the general idea is that we do logical type eq check in logical layer. Do actual casting in physical layer.

So my actual question could be summarised as: what would the "actual casting in physical layer" return? A Datum?

Reply:
`actual casting in physical layer` what in my mind is array with arrow::cast

```rust
let cast_array = kernels::cast::cast_with_options(
                    &scalar_array,
                    cast_type,
                    &cast_options,
                )?;
```

Btw, I'm confused about what is the difference between `ColumnarValue` and `Datum`. It seems like we could replace `ColumnarValue` with `Datum` 🤔

Reply:
> Datum is another concept other than ColumnarValue. It could be Array or Scalar

`Datum` is the arrow-rs concept equivalent of `ColumnarValue::Scalar` (though its implementation is different). Also, it is not yet widely supported in arrow-rs kernels (it is mostly now supported in the core comparison /equality kernels)

`ColumnarValue` and `ScalarValue` both significantly pre-date the introduction of `Datum` to arrow-rs.

Reply:
I've clarified what I mean when discussing `Datum` in this comment:

https://github.com/apache/datafusion/pull/11978#issuecomment-2294834051

Reply:
I think there was a discussion about replacing `ColumnarValue::Scalar` or `ScalarValue` with `Datum`s `Scalar` when it is introduced. But the cost of overhead might not be negligible so we stick with `ScalarValue`

Reply:
https://github.com/apache/datafusion/issues/7353